src/arithmetic/algebraic_expression/algebraic_expression_construction.c

```diff
@@ -0,0 +1,532 @@
+	 * have a variable length edge in them. */
+	for(size_t expIdx = 0; expIdx < expCount; expIdx++) {
+		AlgebraicExpression *exp = expressions[expIdx];
+		if(!_AlgebraicExpression_ContainsVariableLengthEdge(qg, exp)) {
+			res = array_append(res, exp);
+			continue;
+		}
+
+		// Expression contains a variable length edge.
+		QGNode *src = QueryGraph_GetNodeByAlias(qg, AlgebraicExpression_Source(exp));
+
+		// A variable length expression with a labeled source node
+		// We only care about the source label matrix, when it comes to
+		// the first expression, as in the following expressions
+		// src is the destination of the previous expression.
+		if(expIdx == 0 && src->label) {
+			// Remove src node matrix from expression.
+			AlgebraicExpression *op = AlgebraicExpression_RemoveLeftmostNode(&exp);
+			res = array_append(res, op);
+		}
+
+		res = array_append(res, exp);
+
+		// If the expression has a labeled destination, separate it into its own expression.
+		QGNode *dest = QueryGraph_GetNodeByAlias(qg, AlgebraicExpression_Destination(exp));
+		if(dest->label) {
+			// Remove dest node matrix from expression.
+			AlgebraicExpression *op = AlgebraicExpression_RemoveRightmostNode(&exp);
+
+			/* See if dest mat can be prepended to the following expression.
+			 * If not create a new expression. */
+			if(expIdx < expCount - 1 &&
+			   !_AlgebraicExpression_ContainsVariableLengthEdge(qg, expressions[expIdx + 1])) {
+				expressions[expIdx + 1] = _AlgebraicExpression_MultiplyToTheLeft(op, expressions[expIdx + 1]);
+			} else {
+				res = array_append(res, op);
+			}
+		}
+	}
+
+	array_free(expressions);
+	return res;
+}
+
+/* Break down path into sub paths.
+ * Considering referenced intermidate nodes and edges. */
+static QGEdge ***_Intermediate_Paths
+(
+	QGEdge **path,
+	const QueryGraph *qg
+) {
+	QGEdge *e = NULL;
+	int pathLen = array_len(path);
+
+	/* Allocating maximum number of expression possible. */
+	QGEdge ***paths = array_new(QGEdge **, pathLen);
+	QGEdge **intermediate_path = array_new(QGEdge *, pathLen);
+	paths = array_append(paths, intermediate_path);
+
+	/* Scan path left to right,
+	 * construct intermidate paths by "breaking" on referenced entities. */
+	for(int i = 0; i < pathLen - 1; i++) {
+		e = path[i];
+		intermediate_path = array_append(intermediate_path, e);
+		if(_should_divide_expression(path, i, qg)) {
+			// Break! add current path to paths and create a new path.
+			intermediate_path = array_new(QGEdge *, pathLen);
+			paths = array_append(paths, intermediate_path);
+		}
+	}
+
+	// Handle last hop.
+	e = path[pathLen - 1];
+	intermediate_path = array_append(intermediate_path, e);
+
+	return paths;
+}
+
+static AlgebraicExpression *_AlgebraicExpression_OperandFromNode
+(
+	QGNode *n
+) {
+	GrB_Matrix mat;
+	bool free = false;
+	bool diagonal = true;
+	bool transpose = false;
+	Graph *g = QueryCtx_GetGraph();
+
+	if(n->labelID == GRAPH_UNKNOWN_LABEL) mat = Graph_GetZeroMatrix(g);
```

Comment:
I'm rather curious about the possibility of using a magic value like `IDENTITY_MATRIX` in place of the zero matrix, but it would require conditional handing in all `_Eval_*` routines - possibly not worth it.

Reply:
Agree, especially due to the fact that nnz(zero) = 0